### ISSUE TITLE
InvalidRange or EOFError opening previous versions of gzipped files

### ISSUE THREAD

**Issue by ajkerrigan** [ajkerrigan/visidata-plugins#12]
With version support enabled, I've found an issue that seems to only affect _some_ previous versions of gzipped objects. I can open current versions fine, and previous versions of _uncompressed_ files fine. But for some previous versions of gzipped files, I consistently get either an `InvalidRange` error from botocore or an `EOFError` from gzip (full tracebacks below).

These issues are reproducible outside VisiData, so it seems like something related to either s3fs or fsspec. Possibly related to dask/s3fs#225? I can work around some of the errors by setting `cache_type` to `'none'` and/or messing with `block_size` during the s3fs `open()` call, but it's not 100%. Putting a flag here so I can dig more, and so I have a reference in case anyone else bumps into the same trouble.

Version info:
* Python: 3.8.5
* s3fs: 0.5.1
* fsspec: 0.8.3

**InvalidRange** traceback:

```
Traceback (most recent call last):
  File "/Users/akerrigan/.local/pipx/venvs/visidata/lib/python3.8/site-packages/s3fs/core.py", line 207, in _call_s3
    return await method(**additional_kwargs)
  File "/Users/akerrigan/.local/pipx/venvs/visidata/lib/python3.8/site-packages/aiobotocore/client.py", line 151, in _make_api_call
    raise error_class(parsed_response, operation_name)
botocore.exceptions.ClientError: An error occurred (InvalidRange) when calling the GetObject operation: The requested range is not satisfiable

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/akerrigan/.local/pipx/venvs/visidata/lib/python3.8/site-packages/visidata/threads.py", line 207, in _toplevelTryFunc
    t.status = func(*args, **kwargs)
  File "/Users/akerrigan/.local/pipx/venvs/visidata/lib/python3.8/site-packages/visidata/sheets.py", line 267, in reload
    for r in self.iterload():
  File "/Users/akerrigan/.local/pipx/venvs/visidata/lib/python3.8/site-packages/visidata/loaders/json.py", line 28, in iterload
    ret = json.load(fp, object_pairs_hook=OrderedDict)
  File "/usr/local/Cellar/python@3.8/3.8.5/Frameworks/Python.framework/Versions/3.8/lib/python3.8/json/__init__.py", line 293, in load
    return loads(fp.read(),
  File "/usr/local/Cellar/python@3.8/3.8.5/Frameworks/Python.framework/Versions/3.8/lib/python3.8/gzip.py", line 292, in read
    return self._buffer.read(size)
  File "/usr/local/Cellar/python@3.8/3.8.5/Frameworks/Python.framework/Versions/3.8/lib/python3.8/gzip.py", line 485, in read
    buf = self._fp.read(io.DEFAULT_BUFFER_SIZE)
  File "/usr/local/Cellar/python@3.8/3.8.5/Frameworks/Python.framework/Versions/3.8/lib/python3.8/gzip.py", line 87, in read
    return self.file.read(size)
  File "/Users/akerrigan/.local/pipx/venvs/visidata/lib/python3.8/site-packages/fsspec/spec.py", line 1388, in read
    out = self.cache._fetch(self.loc, self.loc + length)
  File "/Users/akerrigan/.local/pipx/venvs/visidata/lib/python3.8/site-packages/fsspec/caching.py", line 350, in _fetch
    new = self.fetcher(self.end, bend)
  File "/Users/akerrigan/.local/pipx/venvs/visidata/lib/python3.8/site-packages/s3fs/core.py", line 1425, in _fetch_range
    return _fetch_range(self.fs, self.bucket, self.key, self.version_id,
  File "/Users/akerrigan/.local/pipx/venvs/visidata/lib/python3.8/site-packages/s3fs/core.py", line 1540, in _fetch_range
    resp = fs.call_s3(fs.s3.get_object, Bucket=bucket, Key=key,
  File "/Users/akerrigan/.local/pipx/venvs/visidata/lib/python3.8/site-packages/fsspec/asyn.py", line 118, in wrapper
    return maybe_sync(func, self, *args, **kwargs)
  File "/Users/akerrigan/.local/pipx/venvs/visidata/lib/python3.8/site-packages/fsspec/asyn.py", line 97, in maybe_sync
    return sync(loop, func, *args, **kwargs)
  File "/Users/akerrigan/.local/pipx/venvs/visidata/lib/python3.8/site-packages/fsspec/asyn.py", line 68, in sync
    raise exc.with_traceback(tb)
  File "/Users/akerrigan/.local/pipx/venvs/visidata/lib/python3.8/site-packages/fsspec/asyn.py", line 52, in f
    result[0] = await future
  File "/Users/akerrigan/.local/pipx/venvs/visidata/lib/python3.8/site-packages/s3fs/core.py", line 225, in _call_s3
    raise translate_boto_error(err) from err
OSError: [Errno 22] The requested range is not satisfiable
```

**EOFError** traceback:

```
Traceback (most recent call last):
  File "/Users/akerrigan/.local/pipx/venvs/visidata/lib/python3.8/site-packages/visidata/threads.py", line 207, in _toplevelTryFunc
    t.status = func(*args, **kwargs)
  File "/Users/akerrigan/.local/pipx/venvs/visidata/lib/python3.8/site-packages/visidata/sheets.py", line 267, in reload
    for r in self.iterload():
  File "/Users/akerrigan/.local/pipx/venvs/visidata/lib/python3.8/site-packages/visidata/loaders/json.py", line 28, in iterload
    ret = json.load(fp, object_pairs_hook=OrderedDict)
  File "/usr/local/Cellar/python@3.8/3.8.5/Frameworks/Python.framework/Versions/3.8/lib/python3.8/json/__init__.py", line 293, in load
    return loads(fp.read(),
  File "/usr/local/Cellar/python@3.8/3.8.5/Frameworks/Python.framework/Versions/3.8/lib/python3.8/gzip.py", line 292, in read
    return self._buffer.read(size)
  File "/usr/local/Cellar/python@3.8/3.8.5/Frameworks/Python.framework/Versions/3.8/lib/python3.8/gzip.py", line 470, in read
    self._read_eof()
  File "/usr/local/Cellar/python@3.8/3.8.5/Frameworks/Python.framework/Versions/3.8/lib/python3.8/gzip.py", line 514, in _read_eof
    crc32, isize = struct.unpack("<II", self._read_exact(8))
  File "/usr/local/Cellar/python@3.8/3.8.5/Frameworks/Python.framework/Versions/3.8/lib/python3.8/gzip.py", line 416, in _read_exact
    raise EOFError("Compressed file ended before the "
EOFError: Compressed file ended before the end-of-stream marker was reached
```

**Comment by ajkerrigan** [comment]
I _think_ I've tracked this down, and reported my suspicions in dask/s3fs#381. I was able to make some local tweaks to s3fs and fsspec to get this working, but I have 0 confidence that they're good changes. I'll let people who know things make the call about what if anything actually needs fixing upstream. It may be that I'm doing silly things I shouldn't be doing in the first place.

In the short term I've worked around the issue in fcfecbbccdc9ca8b7a87cf6c13a5276d0cad7584 by adding a check to be sure an object's cache size matches the target version size. If I _am_ doing something wrong in the first place, this hack is doubling down on that.
